From e2bf8f3a3e81f1aa94fd15fa206dd4ec5ddf2ec7 Mon Sep 17 00:00:00 2001
From: Frediano Ziglio <freddy77@gmail.com>
Date: Sat, 14 Feb 2026 17:16:58 +0000
Subject: [PATCH] utils: Fix compatibility with Nettle 4 library

Hash length parameter was removed from nettle_md5_digest.

Method suggested by Satadru Pramanik.

Signed-off-by: Frediano Ziglio <freddy77@gmail.com>
---
 include/freetds/utils/md5.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/include/freetds/utils/md5.h b/include/freetds/utils/md5.h
index c134b0c49c..ff59a936c6 100644
--- a/include/freetds/utils/md5.h
+++ b/include/freetds/utils/md5.h
@@ -25,6 +25,7 @@ typedef struct MD5Context MD5_CTX;
 #else
 
 #include <nettle/md5.h>
+#include <nettle/version.h>
 
 typedef struct md5_ctx MD5_CTX;
 
@@ -40,7 +41,11 @@ static inline void MD5Update(MD5_CTX *ctx, const uint8_t *buf, size_t len)
 
 static inline void MD5Final(MD5_CTX *ctx, uint8_t *digest)
 {
+#if defined(NETTLE_VERSION_MAJOR) && NETTLE_VERSION_MAJOR >= 4
+	nettle_md5_digest(ctx, digest);
+#else
 	nettle_md5_digest(ctx, 16, digest);
+#endif
 }
 
 #endif
From 67663a02c5ba746254fadb882ff22798c2590d8d Mon Sep 17 00:00:00 2001
From: Frediano Ziglio <freddy77@gmail.com>
Date: Fri, 15 May 2026 22:38:43 +0100
Subject: [PATCH] Fix compatibility with Nettle 4 library

Hash length parameter was removed from digest functions like
nettle_md4_digest and sha1_digest.

Change similar to commit e2bf8f3a3e81f1aa94fd15fa206dd4ec5ddf2ec7
(cfr "utils: Fix compatibility with Nettle 4 library").

Signed-off-by: Frediano Ziglio <freddy77@gmail.com>
---
 include/freetds/utils/md4.h    | 5 +++++
 src/tds/sec_negotiate_gnutls.h | 5 +++++
 2 files changed, 10 insertions(+)

diff --git a/include/freetds/utils/md4.h b/include/freetds/utils/md4.h
index 0a4e3abbbf..61109d9186 100644
--- a/include/freetds/utils/md4.h
+++ b/include/freetds/utils/md4.h
@@ -23,6 +23,7 @@ typedef struct MD4Context MD4_CTX;
 #else
 
 #include <nettle/md4.h>
+#include <nettle/version.h>
 
 typedef struct md4_ctx MD4_CTX;
 
@@ -38,7 +39,11 @@ static inline void MD4Update(MD4_CTX *ctx, const uint8_t *buf, size_t len)
 
 static inline void MD4Final(MD4_CTX *ctx, uint8_t *digest)
 {
+#if defined(NETTLE_VERSION_MAJOR) && NETTLE_VERSION_MAJOR >= 4
+	nettle_md4_digest(ctx, digest);
+#else
 	nettle_md4_digest(ctx, 16, digest);
+#endif
 }
 
 
diff --git a/src/tds/sec_negotiate_gnutls.h b/src/tds/sec_negotiate_gnutls.h
index 10bf90b2fc..6a1f8d5218 100644
--- a/src/tds/sec_negotiate_gnutls.h
+++ b/src/tds/sec_negotiate_gnutls.h
@@ -35,6 +35,7 @@
 #  include <nettle/asn1.h>
 #  include <nettle/rsa.h>
 #  include <nettle/bignum.h>
+#  include <nettle/version.h>
 #endif
 
 /**
@@ -186,7 +187,11 @@ sha1(uint8_t *hash, const void *data, size_t len)
 	struct sha1_ctx ctx;
 	sha1_init(&ctx);
 	sha1_update(&ctx, len, (const uint8_t *) data);
+#if defined(NETTLE_VERSION_MAJOR) && NETTLE_VERSION_MAJOR >= 4
+	sha1_digest(&ctx, hash);
+#else
 	sha1_digest(&ctx, 20, hash);
+#endif
 }
 #endif
 
